### ISSUE TITLE
'each' definition incomplete

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#894]
**Reported by choeger on 1 Nov 2012 10:39 UTC**
The definition of `each` in section 7.2.5 implicitly requires the modifier to be an array (thus it can be splitted and applied). 

This definition leaves some crucial points out:

 1. what if the modifier is not an array
   a. what if the modifier is a redeclaration
 1. what if the modified element is not an array
 1. what if the modified elements and the modifiers dimensions don't match
  a. can a matrix modify a vector
  a. can a vector modify a matrix
 1. what if the modified element and the modifiers length don't match
  a. if the modifier is smaller
  a. if the modifier is larger
   
Point 4 seems to be implicitly defined as "don't modify". Point 3 might be covered by the last sentence of the first bullet point, but this sentence might also refer to the ordering mentioned before.

The more interesting points are 1. and 2. Here is an example:

```mo
class EachTest
  class Foo
    replaceable type S = Integer;
    S x;
  end Foo;

  Foo[1] foo1(redeclare type S = Real);
  Foo[1] foo2(redeclare each type S = Real);
  Foo foo3(redeclare each type S = Real);
end EachTest;
```

Should this cause errors?

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/894

**Comment by modelica-trac-importer** [comment]
**Modified by choeger on 1 Nov 2012 10:40 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#46;se on 1 Nov 2012 11:12 UTC**
1,2) I see nothing in that text that requires the modification to be an array. It requires the declaration to be an array. But the text could be made clearer (e.g. by removing /modification in declaration/modification).

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 1 Nov 2012 11:18 UTC**
Section 7.2.5 is about modifiers for arrays and should be cleaned up a lot; e.g. it should not start by defining 'each'.


As I see it the basic ideas are:
1. Modifiers applied as modifiers for arrays are (except when the modifier directly has each) split to the corresponding array elements.

What should be clarified: 
* It should be made clear that the modified array and the modifier should be the same dimensionality and size. Just ignoring left-over elements is not the Modelica-style (see section 6.6).
* For redeclarations this currently never makes sense; but currently it is only an error if the modifier is nested. It should be an error in all cases.
* Can the modifier implictly define the array size if the array size was declared with ":"?: I would say this should only and explicitly be allowed for the start-attribute (haven't checked if this is already written down).

2. Each can only be applied to modifiers for arrays; and changes the treatment of the modifier so that the modifier (without each) is applied to each element of the array. 

"Modifiers for arrays" might be a bit unclear for the spec; the idea is that the modified element is an array; and the modifier can be direct in the declaration or nested later.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 1 Nov 2012 11:37 UTC**
Replying to [comment:2 sjoelund.se]:
> 1,2) I see nothing in that text that requires the modification to be an array. It requires the declaration to be an array. But the text could be made clearer (e.g. by removing /modification in declaration/modification).

I agree with the first parts and that the text in 7.2.5 is not the best as previously indicated. However, "/modification" cannot be removed since this applies for all modifiers - even the ones in extends-clauses.

To be clear: yes, 7.2.5 implicitly requires the modification to be an array; in the case of an array and a modifier **without** 'each'. With 'each' it implicitly requires the modifier to be a scalar (since it is applied to each element of the array).

As for the example: foo2 is legal (except for redeclaring Integer to Real); the 'each' in foo3 is explicitly forbidden by 7.2.5 (no array!).

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 1 Nov 2012 12:06 UTC**
Replying to [comment:4 hansolsson]:
> Replying to [comment:2 sjoelund.se]:
> > 1,2) I see nothing in that text that requires the modification to be an array. It requires the declaration to be an array. But the text could be made clearer (e.g. by removing /modification in declaration/modification).

As Hans already pointed out, that was meant for the case `each` is absent.

 
> To be clear: yes, 7.2.5 implicitly requires the modification to be an array; in the case of an array and a modifier **without** 'each'. With 'each' it implicitly requires the modifier to be a scalar (since it is applied to each element of the array).

And this should be made explicit IMO. Additionally it should also be made explicit that `each` serves no purpose in case of a type-redeclaration.
 
> As for the example: foo2 is legal (except for redeclaring Integer to Real); the 'each' in foo3 is explicitly forbidden by 7.2.5 (no array!).

I wanted to discuss exactly this: If `each` influences only a value modification, and we are redeclaring a type (and yes, it should have been covariant), why do we forbid it here and allow it at another place? I mean, if it does not actually change the behaviour for `foo2` and is allowed there why forbid it in `foo3`. I am in favour of forbidding it in both cases.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 4 Dec 2012 22:54 UTC**
Language group: Not possible to clean this up for 3.2 rev2; and primarily an issue with clarification.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 28 Feb 2017 17:05 UTC**
Language group: Not possible to clean this up for Modelica 3.4.
Basic idea would be to start with non-each case.

**Comment by HansOlsson** [comment]
This seems to be closed after #2648
